File: cpp_lib/src/trees/prefix_tree.cpp
//
// Created by psycho on 20.12.2021.
//

#include "prefix_tree.h"
#include <stdio.h>

prefix_tree_node::prefix_tree_node() {
  this->value = *new python::object();
  this->terminations = 0;
  this->children = new std::unordered_map<python::object, prefix_tree_node*>();
}

prefix_tree_node::prefix_tree_node(python::object value) : prefix_tree_node() {
  this->value = std::move(value);
}

size_t std::hash<python::object>::operator()(const python::object &obj) const noexcept {
  return PyObject_Hash(obj.ptr());
}

prefix_tree::prefix_tree() {
  this->root = new prefix_tree_node();
}

void prefix_tree::add(python::object &sequence) {
  auto iter = PyObject_GetIter(sequence.ptr());
  if (iter == nullptr){
    return;
  }

  PyObject *item;
  uint i;
  auto node = this->root;
  for (i = 0, item = PyIter_Next(iter); item; ++i, item = PyIter_Next(iter)){
    python::object obj((python::handle<>(item)));
    if (node->children->contains(obj)){
      node = node->children->at(obj);
    }else{
      auto new_node = new prefix_tree_node(obj);
      node->children->insert({obj, new_node});
      node = new_node;
    }
  }
  node->terminations++;
}

prefix_tree_node *prefix_tree::catch_node(python::object &sequence) const{
  auto iter = PyObject_GetIter(sequence.ptr());
  if (iter == nullptr){
    return nullptr;
  }

  PyObject *item;
  uint i;
  auto node = this->root;
  for (i = 0, item = PyIter_Next(iter); item; ++i, item = PyIter_Next(iter)){
    python::object obj((python::handle<>(item)));
    if (node->children->contains(obj)){
      node = node->children->at(obj);
    }else{
      return nullptr;
    }
  }
  return node;
}

bool prefix_tree::contains(python::object &sequence) const {
  return (this->count(sequence) > 0);
}

uint prefix_tree::count(python::object &sequence) const {
  auto node = this->catch_node(sequence);
  if (node == nullptr){
    return 0;
  }
  return node->terminations;
}

File: structures/pytrees/_PrefixTree.py
from typing import Union, Iterable
from .._internal import EqType


class PrefixTreeNode:
    def __init__(self, value: EqType = None):
        self._value = value
        self._termination = 0
        self._children = dict()
        self._child_values = set()

    def add_child(self, value: EqType) -> 'PrefixTreeNode':
        if value not in self._child_values:
            self._children[value] = (node := PrefixTreeNode(value))
            self._child_values.add(value)
        else:
            node = self.get_child(value)

        return node

    def get_child(self, value: EqType) -> 'PrefixTreeNode':
        if value in self._child_values:
            return self._children[value]
        else:
            return None

    def terminate(self):
        self._termination += 1

    @property
    def value(self) -> str:
        return self._value

    @property
    def termination(self):
        return self._termination

    def __contains__(self, item: 'PrefixTreeNode') -> bool:
        return item in self._child_values

    def __bool__(self) -> bool:
        return self.value is not None

    def __str__(self) -> str:
        return f"PrefixTreeNode[{self.value}]"

    def __eq__(self, other: 'PrefixTreeNode') -> bool:
        return self.value == other.value


class PrefixTree:
    def __init__(self):
        self._root = PrefixTreeNode()

    def add(self, sequence: Union[EqType, Iterable[EqType]]):
        if hasattr(sequence, '__iter__'):
            current_node = self._root
            for v in sequence:
                current_node = current_node.add_child(v)
            current_node.terminate()
        else:
            self._root.add_child(sequence)

    def count(self, sequence: Iterable[EqType]) -> int:
        current_node = self._root
        for v in sequence:
            if (current_node := current_node.get_child(v)) is None:
                return False
        return current_node.termination

    def contains(self, sequence: Iterable[EqType]) -> bool:
        return sequence in self

    def __contains__(self, sequence: Iterable[EqType]) -> bool:
        return self.count(sequence) > 0
